src/serializer.rs

```diff
@@ -90,90 +61,277 @@ impl<'a> Message<'a> {
                     message_type: MessageType::Notification,
                     return_code: ReturnCode::Ok,
                 };
-                MAGIC_COOKIE.to_writer(writer).map(|_| 16)
+                MAGIC_COOKIE.to_writer(writer).map(|_| HEADER_LENGTH)
             }
         }
     }
 }
 
-/// Serialization of a SomeIP message to bytes
-#[deprecated(note = "use SomeIp::to_vec instead")]
-#[allow(dead_code)]
-pub fn serialize_someip(someip: &Message) -> Vec<u8> {
-    someip.to_vec()
+impl Header {
+    pub fn to_vec(&self) -> Vec<u8> {
+        let mut buf = Vec::with_capacity(HEADER_LENGTH);
+        self.to_writer(&mut buf).unwrap(); // Safe because it is a Vec
+        buf
+    }
+
+    pub fn to_writer<W: Write>(&self, mut writer: W) -> Result<usize, Error> {
+        writer.write_u16::<BigEndian>(self.message_id.service_id)?;
+        writer.write_u16::<BigEndian>(self.message_id.method_id)?;
+        writer.write_u32::<BigEndian>(self.length)?;
+        writer.write_u16::<BigEndian>(self.request_id.client_id)?;
+        writer.write_u16::<BigEndian>(self.request_id.session_id)?;
+        writer.write_u8(self.protocol_version)?;
+        writer.write_u8(self.interface_version)?;
+        writer.write_u8(self.message_type.into())?;
+        writer.write_u8(self.return_code.into())?;
+        Ok(HEADER_LENGTH)
+    }
+}
+
+impl SdPayload {
+    pub fn to_vec(&self) -> Vec<u8> {
+        let mut buf = Vec::with_capacity(self.length());
+        self.to_writer(&mut buf).unwrap(); // Safe because it is a Vec
+        buf
+    }
+
+    pub fn to_writer<W: Write>(&self, mut writer: W) -> Result<usize, Error> {
+        writer.write_u8(self.flags)?;
+        writer.write_u24::<BigEndian>(0x000000)?; // reserved
+
+        writer.write_u32::<BigEndian>(self.entries_length() as u32)?;
+        for entry in &self.entries {
+            let entry_type = entry.get_type();
+            match entry {
+                SdEntry::FindService(item) => item.to_writer(entry_type, &mut writer)?,
+                SdEntry::OfferService(item) => item.to_writer(entry_type, &mut writer)?,
+                SdEntry::SubscribeEventgroup(item) => item.to_writer(entry_type, &mut writer)?,
+                SdEntry::SubscribeEventgroupAck(item) => item.to_writer(entry_type, &mut writer)?,
```

Comment:
This is four times the same? Maybe add a a SdEntry::to_writer to encapsulate it at least.